# 2. teden: Numpy - Domača naloga

Uvozimo modul numpy:

In [1]:
import numpy as np

### 1.Create a null vector of size 10

In [3]:
a = np.zeros(10)
print(a)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [8]:
b = np.zeros(10)
b[4]=1
print(b)

[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [10]:
c = np.arange(10,50)
print(c)

[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]


Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [11]:
Z = np.arange(50)

In [14]:
Z = np.flip(Z)
print(Z)

[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26
 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2
  1  0]


Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [16]:
d = np.arange(9).reshape(3,3)
print(d)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [17]:
a = [1,2,0,0,4,0]

In [26]:
b = np.where(a!=0)

In [27]:
print(b)

(array([0, 1, 4], dtype=int32),)


Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [30]:
e = np.identity(3)
print(e)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [32]:
f = np.random.random((3,3,3))
print(f)

[[[0.17771791 0.90945859 0.71125867]
  [0.87862652 0.70385342 0.27207287]
  [0.78412419 0.80149752 0.49205595]]

 [[0.41552177 0.8295003  0.06026666]
  [0.92277865 0.05781965 0.86332435]
  [0.93631297 0.51667092 0.35171927]]

 [[0.68472838 0.84911675 0.92943944]
  [0.15300634 0.38776587 0.36884776]
  [0.72337315 0.2643057  0.15656198]]]


Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [36]:
g = np.random.random((10,10))
Zmin = g.min()
Zmax = g.max()
print('Zmin='+str(Zmin)+' Zmax='+str(Zmax))

Zmin=0.0015942025254671588 Zmax=0.9991577438648033


Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [38]:
h = np.ones((10,10))
h[1:-1,1:-1] = 0
print(h)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

In [39]:
0 * np.nan

nan

In [40]:
np.nan == np.nan

False

In [41]:
np.inf > np.nan

False

In [42]:
np.nan - np.nan

nan

In [43]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [52]:
ch = np.zeros((8,8), dtype=int)
ch[1::2,::2] = 1
ch[::2,1::2] = 1

In [53]:
print(ch)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [73]:
Z = np.arange(11)

In [74]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [75]:
Z = np.where((Z>=4) & (Z<=8), Z*-1, Z)

In [76]:
print(Z)

[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]


Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [81]:
i = np.linspace(0,1,12,endpoint=True)[1:-1]
print(i)

[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455
 0.63636364 0.72727273 0.81818182 0.90909091]


Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [85]:
j = np.sort(np.random.random(10))
print(j)

[0.03561791 0.17736065 0.25524848 0.33463762 0.37903428 0.48961128
 0.50730096 0.62990572 0.91405875 0.97368709]


Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [86]:
A = np.random.randint(0,2,5)

In [87]:
B = np.random.randint(0,2,5)

In [91]:
np.array_equal(A, B)

True

### 17.Make an array immutable (read-only)

In [92]:
Z = np.zeros(10)

In [93]:
Z.flags.writeable = False

In [94]:
Z[0] = '0.2'

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [106]:
r = np.random.random(10)
print(r)

[0.67821288 0.81219041 0.27175414 0.89769794 0.17801359 0.09659442
 0.95463107 0.49324723 0.37402298 0.40239488]


In [107]:
r[r.argmax()] = 0
print(r)

[0.67821288 0.81219041 0.27175414 0.89769794 0.17801359 0.09659442
 0.         0.49324723 0.37402298 0.40239488]


### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [108]:
Z = np.arange(100)
print(Z)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]


In [109]:
v = np.random.uniform(0,100)
print(v)

45.30511207619048


In [112]:
ind = (np.abs(Z - v)).argmin()
print(Z[ind])

45


### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [113]:
Z = np.arange(10, dtype=np.int32)

In [114]:
print(Z)

[0 1 2 3 4 5 6 7 8 9]


In [117]:
Z = np.array(Z, dtype ='float32' )

In [118]:
print(Z)

[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]


Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [119]:
Z = np.random.randint(0,10,50)
print(Z)

[7 5 7 7 2 9 7 4 7 4 2 5 6 1 7 2 8 8 3 9 1 3 6 1 0 8 0 8 3 6 2 6 6 6 7 0 2
 1 4 0 9 1 8 6 8 2 3 0 8 2]


In [121]:
st = np.bincount(Z)
print(st)

[5 5 7 4 3 2 7 7 7 3]


In [127]:
#print(st==np.max(st))

[False False  True False False False  True  True  True False]


In [128]:
st2 = np.where(st==np.max(st))

In [135]:
print(np.asarray(st2))

[[2 6 7 8]]
